### ISSUE TITLE
Keyspotting : using localstorage to save the training set ?

### ISSUE THREAD

**Issue by fritz-smh** [dreamdom/JsSpeechRecognizer#5]
Hi,

I did some tests with the keyspotting feature and it is really nice !
I really would like to implement this in the web client of Domogik, an open source home automation solution.
Domogik has its own virtual assistant, as the famous Jarvis to do various operations : control the home, discuss a little, learn some things, ... Example with the android client : https://www.youtube.com/watch?v=IXXahef0bNY

In the web application, I implemented TTS and STT feature, but the keyspotting feature still needs to be added and your project seems to be the solution for me.

By the way, for now, we need to train each time we reload the page. Do you have any plan to use some localstorage in your library to store the training set ?

Once again, goog job!

++
Fritz

**Comment by dreamdom** [comment]
I'm glad to hear you like the project. I watched the video of your Android client and it looks very interesting. With some help from Google translate I could see that you were able to teach your app to recognize new phrases. Very cool!

And yes, I will be looking at some ways to save the data in between page reloads. I have not used LocalStorage before, but I will be looking into it.

**Comment by Pierre-Gilles** [comment]
It's really simple, you should just provide two methods : 
- A `toJson()` method which returns a `JSON.stringify` of your variable storing the trained set. ( Then the developer of the app should simply handle the "save" part and save it wherever he wants ( on the localStorage, or server-side so the set can be loaded on every client) )
- A `load(string)` method taking in parameters the string. The method should simply parse the JSON and affect it to the same variable your stringified before.

**Comment by Nixellion** [comment]
Yes, this sounds like a crucial thing. I can't think of any real-life use for this if you have to train it new words every time you reload the page, or restart it somehow. This is first thing that has to be done.

Another thing that I would love to see, is the ability to stop and release the microphone, unless I can already do that. I want to start google cloud speech recognition after the hotword was detected. Obviously - similar to Ok Google, Hey Cortana and.. wahtever siri responds to :D and Alexa. Because in some (if not all) cases you can't use 2 streams of audio, so you have to release the microphone before starting speech recognizer. Then release it again and restart keyword detection.

**Comment by Nixellion** [comment]
So, any ETA on this feature? It would really allow us to start using it not as a prototype, but as a production.. thing  :D 

Because training a keyword each time you reload your voice assistant is just not right haha :D 

I'm asking to know if I have to start digging into it myself, and trying to write my own storage procedures to your system or looking for another approach entirely, or if I should just wait for it.

Thanks,
